File: 49-group-anagrams/group-anagrams.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function(strs) {
    let map = {}; // This will group words by their sorted letters

    for (let word of strs) {
        // Sort the letters in the word to create a key
        let sorted = word.split('').sort().join('');

        // If key doesn't exist, create it with an empty array
        if (!map[sorted]) {
            map[sorted] = [];
        }

        // Push the word into the group
        map[sorted].push(word);
    }

    // Return all grouped anagrams as an array
    return Object.values(map);
};

File: 49-group-anagrams/group-anagrams.py
class Solution:
    from typing import List

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        map = {}  # Dictionary to group anagrams

        for word in strs:
            # Sort the word to create a key
            sorted_word = ''.join(sorted(word))

            # If the key is not in the map, create it
            if sorted_word not in map:
                map[sorted_word] = []

            # Add the original word to the list
            map[sorted_word].append(word)

        # Return all grouped anagrams
        return list(map.values())